### PR TITLE
ManagedInstance api sdk implimentation

### PR DESCRIPTION
**Pull request by ramrutha497**
Checklist
---------

JIRA Ticket -

Checklist (Please mark Yes/No/NA for each of the questions below) -

- I have provided clear and detailed explanation of changes made in the proposed PR : No
- I have compiled the code successfully my laptop : No
- I have written examples to demonstrates the usage of the newly implemented APIs : No
- I have incremented the release version : No
- I have run the Intellij Code Analysis on newly added code and fixed all (valid) reported issues : No

OR

- I have taken approval from scrum master or project manager for skipping any of the items above : No

### CODE REVIEW

File: src/main/java/com/spotinst/sdkjava/model/ISpotAwsManagedInstanceRepo.java [hunk 1]
```diff
@@ -39,4 +39,12 @@ default RepoGenericResponse<GetMigrationStatus> getMigrationStatus(String manage
     default RepoGenericResponse<List<GetAllManagedInstancesResponse>> getAllManagedInstances(String authToken, String account){
         throw new SpotinstNotSupportedException();
     }
+
+    RepoGenericResponse<GetInstanceCost> getInstanceCosts(String migrationId, String fromDate, String toDate, String authToken, String account);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/ISpotAwsManagedInstanceRepo.java:43]: first parameter should be managedInstanceId

File: src/main/java/com/spotinst/sdkjava/model/SpotAwsManagedInstanceClient.java [hunk 2]
```diff
@@ -275,4 +276,68 @@ public GetMigrationStatus getManagedInstanceMigrationStatus(String managedInstan
         return migrationStatus;
     }
 
+    public GetInstanceCost getManagedInstanceCosts(String managedInstanceId, String fromDate, String toDate) {
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/SpotAwsManagedInstanceClient.java:279]: "aggregationPeriod" query param is missing

File: src/main/java/com/spotinst/sdkjava/model/service/aws/managedInstance/AwsManagedInstanceService.java [hunk 3]
```diff
@@ -454,4 +454,125 @@ public static ApiGetMigrationStatus getManagedInstanceMigrationStatus(String man
 
         return getMigrationStatus;
     }
+
+    public static ApiGetInstanceCost getManagedInstanceCosts(String managedInstanceId, String fromDate, String toDate, String authToken,
+                                                                          String account) throws SpotinstHttpException {
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/service/aws/managedInstance/AwsManagedInstanceService.java:459]: "aggregationPeriod" query param is missing

File: src/main/java/com/spotinst/sdkjava/model/api/aws/managedInstance/ApiRunning.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package com.spotinst.sdkjava.model.api.aws.managedInstance;
+
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/api/aws/managedInstance/ApiRunning.java:2]: re-utilize this class, as ApiSavings class also contains same attributes

File: src/main/java/com/spotinst/sdkjava/model/api/aws/managedInstance/ApiSavings.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
+package com.spotinst.sdkjava.model.api.aws.managedInstance;
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/api/aws/managedInstance/ApiSavings.java:1]: no need of this class, ApiRunning class can be re-utilized

File: src/main/java/com/spotinst/sdkjava/model/bl/aws/managedInstance/Savings.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
+package com.spotinst.sdkjava.model.bl.aws.managedInstance;
+
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/bl/aws/managedInstance/Savings.java:2]: same attributes are available in Running class, re-utilize the same class

File: src/main/java/com/spotinst/sdkjava/model/converters/aws/managedInstance/AwsManagedInstanceConverter.java [hunk 7]
```diff
@@ -1698,6 +1714,30 @@ public static GetMigrationStatus toBl(ApiGetMigrationStatus getMigrationStatus)
...
+            if (getInstanceCosts.isSavingsSet()) {
+                getStatusBuilder.setSavings(getInstanceCosts.getSavings());
+            }
+
+            retVal = getStatusBuilder.build();
+        }
+
+        return retVal;
+    }
+
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/converters/aws/managedInstance/AwsManagedInstanceConverter.java:1740]: add running class, cost class conversion as well

File: src/main/java/com/spotinst/sdkjava/model/repo/aws/managedInstance/SpotAwsManagedInstanceRepo.java [hunk 8]
```diff
@@ -208,5 +208,57 @@ public RepoGenericResponse<GetMigrationStatus> getMigrationStatus(String migrati
         return migrationStatus;
     }
 
+    @Override
+    public RepoGenericResponse<GetInstanceCost> getInstanceCosts(String migrationId, String fromDate, String toDate, String authToken, String account) {
+        RepoGenericResponse<GetInstanceCost> instanceCosts;
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/repo/aws/managedInstance/SpotAwsManagedInstanceRepo.java:213]: change first parameter to managedInstanceId

File: src/main/java/com/spotinst/sdkjava/model/requests/aws/managedInstance/AwsManagedInstanceUpdateRequest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            return managedInstanceUpdateRequest;
+        }
+
+    }
+
+    public String toJson() {
+        ApiManagedInstanceUpdate apiManagedInstanceUpdate = AwsManagedInstanceConverter.toDal(managedInstanceUpdate);
+
+        Map<String, ApiManagedInstanceUpdate> managedInstanceUpdateRequest = new HashMap<>();
+        managedInstanceUpdateRequest.put("managedInstanceUpdate", apiManagedInstanceUpdate);
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/requests/aws/managedInstance/AwsManagedInstanceUpdateRequest.java:55]: "managedInstanceUpdate" should be renamed to "managedInstanceStates"

File: src/main/java/com/spotinst/sdkjava/model/responses/aws/managedInstance/AwsManagedInstanceGetInstanceCostsApiRespponse.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.spotinst.sdkjava.model.responses.aws.managedInstance;
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/responses/aws/managedInstance/AwsManagedInstanceGetInstanceCostsApiRespponse.java:1]: typo in file name

File: src/main/java/com/spotinst/sdkjava/model/service/aws/managedInstance/AwsManagedInstanceService.java [hunk 11]
```diff
@@ -454,4 +454,125 @@ public static ApiGetMigrationStatus getManagedInstanceMigrationStatus(String man
...
+        Map<String, String> headers = buildHeaders(authToken);
+
+        // Build URI
+        String uri = String.format("%s/aws/ec2/managedInstance/%s/costs", apiEndpoint, managedInstanceId);
+
+        // Send the request
+        RestResponse response = RestClient.sendGet(uri, headers, queryParams);
+
+        // Handle the response.
+        AwsManagedInstanceGetInstanceCostsApiRespponse managedInstanceStatusResponse =
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/service/aws/managedInstance/AwsManagedInstanceService.java:494]: typo in name api response name

File: src/main/java/com/spotinst/sdkjava/model/service/aws/managedInstance/AwsManagedInstanceService.java [hunk 12]
```diff
@@ -454,4 +454,125 @@ public static ApiGetMigrationStatus getManagedInstanceMigrationStatus(String man
...
+        String body = JsonMapper.toJson(managedInstanceUpdateRequest);
+
+        // Build URI
+        String uri = String.format("%s/aws/ec2/managedInstance/state", apiEndpoint);
+
+        // Send the request
+        RestResponse response = RestClient.sendPut(uri, body, headers, queryParams);
+
+        // Handle the response.
+        AwsManagedInstanceApiResponse managedInstanceUpdateResponse =
```
Comment (chandra1-n) [inline src/main/java/com/spotinst/sdkjava/model/service/aws/managedInstance/AwsManagedInstanceService.java:569]: it's a boolean response, no need to caste with AwsManagedInstanceApiResponse

**Review by chandra1-n — CHANGES_REQUESTED** [review]
please fix the review comments

**Comment by anuragsharma-123** [comment]
These APIs are implemented and tested in Python-SDK. Closing this PR.

**Pull request closed without merging**